### DISCUSSION TITLE
Allow permission to be determined by an expression

### DISCUSSION THREAD

**Discussion by qc00** [community/community#53915] [Actions]
### Select Topic Area

Product Feedback

### Body

Currently, the [`permissions`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions) key does not interpret expressions.

In a workflow that supports both `push` and `pull_request` events, it is very useful to evaluate dynamically if the write permission can be requested.

Under the current design, the relevant actions, e.g. package upload, can be made to fail gracefully without `write` permission, but the workflow itself cannot.

This results in unnecessary duplication where two workflows differ only by permission request.

**Comment by jsoref** [comment]
The workaround for this is to stick 95% of your content into an action (possibly local) and then have both workflows use the action.

Also note that a workflow can have multiple jobs, so you can (and I [do](https://github.com/check-spelling/spell-check-this/blob/3a35a1cec1e660ab04638cbd5bdd086cf2fcd5c5/.github/workflows/spelling.yml#L108)) have different jobs for the two cases. You could use a third job to share common calculations and use outputs+inputs to glue things together. 

That said. I'd really love this feature (e.g. [to conditionally ask for sarif](https://github.com/check-spelling/spell-check-this/blob/3a35a1cec1e660ab04638cbd5bdd086cf2fcd5c5/.github/workflows/spelling.yml#L62)). I also am not holding my breath.
